framework/tlvf/src/include/tlvf/WSC/AttrList.h

```diff
@@ -0,0 +1,139 @@
+
+    template <class T> std::shared_ptr<T> getAttr() { return this->getClass<T>(); };
+    template <class T> std::shared_ptr<T> addAttr() { return this->addClass<T>(); };
+    bool finalize() { return ClassList::finalize(); };
+    size_t len() { return this->getMessageLength(); };
+    uint8_t *buffer() { return this->getMessageBuff(); };
+
+    // getter functions for common attributes for m1 and m2
+    std::string manufacturer() { return getAttr<cWscAttrManufacturer>()->manufacturer_str(); };
+    std::string model_name() { return getAttr<cWscAttrModelName>()->model_str(); };
+    std::string device_name() { return getAttr<cWscAttrDeviceName>()->device_name_str(); };
+    std::string serial_number() { return getAttr<cWscAttrSerialNumber>()->serial_number_str(); };
+    uint16_t encr_type_flags()
+    {
+        return getAttr<cWscAttrEncryptionTypeFlags>()->encr_type_flags();
+    };
+    uint16_t auth_type_flags()
+    {
+        return getAttr<cWscAttrAuthenticationTypeFlags>()->auth_type_flags();
+    };
+    uint8_t *enrollee_nonce() { return getAttr<cWscAttrEnrolleeNonce>()->nonce(); };
+    uint8_t *public_key() { return getAttr<cWscAttrPublicKey>()->public_key(); };
+    uint16_t rf_bands() { return getAttr<cWscAttrRfBands>()->bands(); };
+    bool valid();
+
+protected:
+    sWscAttrHeader *getNextAttrHdr()
+    {
+        return reinterpret_cast<sWscAttrHeader *>(
+            m_class_vector.empty() ? getMessageBuff() : m_class_vector.back()->getBuffPtr());
+    }
+    uint16_t getNextAttrType() { return ntohs(getNextAttrHdr()->type); };
+    size_t getRemainingBytes()
+    {
+        return m_class_vector.empty() ? getMessageBuffLength()
+                                      : m_class_vector.back()->getBuffRemainingBytes();
+    };
+    virtual bool valid_custom() = 0;
+
+private:
+    static eWscMessageType wsc_msg_type_from_buffer(uint8_t *buff, size_t buff_len);
+};
+
+class m1 : public AttrList {
+public:
+    m1(uint8_t *buff, size_t buff_len, bool parse) : AttrList(buff, buff_len, parse) {}
+    virtual ~m1() = default;
+
+    virtual eWscMessageType type() override { return eWscMessageType::WSC_MSG_TYPE_M1; };
+    bool init(const config &cfg) override;
+    bool valid_custom() override;
+
+    // getters
+    sMacAddr mac_addr() { return getAttr<cWscAttrMac>()->data(); };
+};
+
+class m2 : public AttrList {
+public:
+    m2(uint8_t *buff, size_t buff_len, bool parse) : AttrList(buff, buff_len, parse) {}
+    virtual ~m2() = default;
+
+    virtual eWscMessageType type() override { return eWscMessageType::WSC_MSG_TYPE_M2; };
+    bool init(const config &cfg) override;
+    bool valid_custom() override;
+
+    // getters
+    uint8_t *authenticator() { return getAttr<cWscAttrAuthenticator>()->data(); };
+    uint8_t *registrar_nonce() { return getAttr<cWscAttrRegistrarNonce>()->nonce(); };
+    cWscAttrEncryptedSettings &encrypted_settings()
+    {
+        return *getAttr<cWscAttrEncryptedSettings>();
```

Comment:
What if this attribute doesn't exists? You'll dereference a nullptr?

Reply:
The design of these classes is that the `AttrList` is an abstract class with factory methods to create and parse M1 and M2 messages. The `AttrList::create()` uses `wsc_msg_type_from_buffer()` to get the message type and then creates a `shared_ptr` of the appropriate derived class - M1 or M2.
Then it calls one of the 2 versions of the `init()` virtual function which calls `addAttr<T>` for all attributes which belong in this message.
Finally, the `valid()` function validates that the message includes all the mandatory attributes, for those we have the getter functions.
Bottom line - M2 and M1 should __only__ be created using one of the factory methods - 
 `AttrList::create()` or `AttrList::parse()`, which ensures all the attributes are there and can be dereferenced. This makes the rest of the code much simpler, especially for the callers which remain almost identical to how they were before this change.

Reply:
ACK